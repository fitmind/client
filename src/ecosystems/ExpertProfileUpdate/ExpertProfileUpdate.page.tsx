import { Formik } from 'formik';
import React from 'react';
import { Button, Card, Col, Container, Form, Image, Row } from 'react-bootstrap';
import { connect } from 'react-redux';
import { RouteComponentProps, withRouter } from 'react-router';
import styled from 'styled-components';
import * as Yup from 'yup';
import CONFIG from '../../config/config';
import { expertProfileUpdateAction, fetchExpertUserAction } from '../../redux/actions/server/server.actions';
import { ApplicationState, ConnectedReduxProps } from '../../redux/reducers/root.reducer';
import { expertUserInterface } from '../../redux/reducers/server-reducer/server.reducer';

const ExpertProfileUpdateSchema = Yup.object().shape({
    firstName: Yup.string()
        .min(2, 'Too Short!')
        .max(24, 'Too Long!')
        .required('Required'),
    lastName: Yup.string()
        .min(2, 'Too Short!')
        .max(24, 'Too Long!')
        .required('Required'),
    interestedInExpertiseAreas: Yup.array()
        .of(Yup.string())
        .min(1)
        .required('Required'),
    description: Yup.string()
        .min(6, 'Too Short!')
        .max(700, 'Too Long!')
        .required('Required'),
    phone: Yup.string()
        .matches(
            /^((\\+[1-9]{1,4}[ \\-]*)|(\\([0-9]{2,3}\\)[ \\-]*)|([0-9]{2,4})[ \\-]*)*?[0-9]{3,4}?[ \\-]*[0-9]{3,4}?$/,
            'Invalid',
        )
        .required('Required'),
});
interface PropsFromState {
    expertUser: expertUserInterface;
}

interface PropsFromDispatch {
    fetchExpertUserAction: typeof fetchExpertUserAction;
    expertProfileUpdateAction: typeof expertProfileUpdateAction;
}
type ExpertProfilePageAllProps = RouteComponentProps & ConnectedReduxProps & PropsFromState & PropsFromDispatch;

export class ExpertProfileUpdate extends React.Component<ExpertProfilePageAllProps> {
    componentDidMount() {
        this.props.fetchExpertUserAction();
    }
    render() {
        const { expertUser } = this.props;
        return (
            <Container fluid>
                <Row>
                    <Col md={4} />
                    <Col md={4}>
                        <CardWrapper>
                            <Card>
                                <Card.Header as="h5">Edit your Profile</Card.Header>
                                <Card.Body>
                                    <CenterContainer>
                                        <Image src={expertUser.pictureUrl} rounded />
                                        <Formik
                                            enableReinitialize={true} // #DOUBT without this, details not coming
                                            initialValues={expertUser}
                                            validationSchema={ExpertProfileUpdateSchema}
                                            onSubmit={(values, { setSubmitting }) => {
                                                this.props.expertProfileUpdateAction(values);
                                                setSubmitting(false);
                                            }}
                                            render={({
                                                values,
                                                errors,
                                                touched,
                                                handleBlur,
                                                handleChange,
                                                handleSubmit,
                                                isSubmitting,
                                                setFieldValue,
                                            }) => (
                                                <Form noValidate onSubmit={handleSubmit}>
                                                    <Form.Group>
                                                        <Form.Label>First Name</Form.Label>
                                                        <Form.Control
                                                            type="text"
                                                            name="firstName"
                                                            placeholder="your name"
                                                            onChange={handleChange}
                                                            onBlur={handleBlur}
                                                            value={values.firstName}
                                                            isValid={touched.firstName && !errors.firstName}
                                                            isInvalid={!!errors.firstName}
                                                        />
                                                        <Form.Control.Feedback>Looks good!</Form.Control.Feedback>
                                                        <Form.Control.Feedback type="invalid">
                                                            {errors.firstName}
                                                        </Form.Control.Feedback>
                                                    </Form.Group>
                                                    <Form.Group>
                                                        <Form.Label>Last Name</Form.Label>
                                                        <Form.Control
                                                            type="text"
                                                            name="lastName"
                                                            placeholder="your last name"
                                                            onChange={handleChange}
                                                            onBlur={handleBlur}
                                                            value={values.lastName}
                                                            isValid={touched.lastName && !errors.lastName}
                                                            isInvalid={!!errors.lastName}
                                                        />
                                                        <Form.Control.Feedback>Looks good!</Form.Control.Feedback>
                                                        <Form.Control.Feedback type="invalid">
                                                            {errors.lastName}
                                                        </Form.Control.Feedback>
                                                    </Form.Group>
                                                    <Form.Group>
                                                        <Form.Label>Your Description</Form.Label>
                                                        <Form.Control
                                                            as="textarea"
                                                            rows={4}
                                                            type="text"
                                                            name="description"
                                                            onChange={handleChange}
                                                            onBlur={handleBlur}
                                                            value={values.description}
                                                            isValid={touched.description && !errors.description}
                                                            isInvalid={!!errors.description}
                                                        />
                                                        <Form.Control.Feedback>Looks good!</Form.Control.Feedback>
                                                        <Form.Control.Feedback type="invalid">
                                                            {errors.description}
                                                        </Form.Control.Feedback>
                                                    </Form.Group>
                                                    <Form.Group>
                                                        <Form.Label>
                                                            What type of coaching are you looking at?
                                                        </Form.Label>
                                                        <Form.Control
                                                            as="select"
                                                            multiple={true}
                                                            name="expertise"
                                                            onChange={evt =>
                                                                setFieldValue(
                                                                    'expertise',
                                                                    [].slice
                                                                        .call(
                                                                            (evt.target as HTMLSelectElement)
                                                                                .selectedOptions,
                                                                        )
                                                                        .map(option => option.value),
                                                                )
                                                            }
                                                            values={expertUser.expertise}
                                                            onBlur={handleBlur}
                                                            isValid={touched.expertise && !errors.expertise}
                                                            isInvalid={!!errors.expertise}
                                                        >
                                                            {' '}
                                                            {Object.keys(CONFIG.expertises).map((key: string) => (
                                                                <option
                                                                    key={key}
                                                                    value={CONFIG.expertises[key].value}
                                                                    selected={
                                                                        values &&
                                                                        values.expertise &&
                                                                        values.expertise.indexOf(
                                                                            CONFIG.expertises[key].value,
                                                                        ) !== -1
                                                                            ? true
                                                                            : false
                                                                    }
                                                                >
                                                                    {CONFIG.expertises[key].display}
                                                                </option>
                                                            ))}
                                                        </Form.Control>
                                                    </Form.Group>

                                                    <Form.Group>
                                                        <Form.Label>Phone</Form.Label>
                                                        <Form.Control
                                                            type="text"
                                                            name="phone"
                                                            placeholder="Enter phone"
                                                            onChange={handleChange}
                                                            onBlur={handleBlur}
                                                            value={values.phone}
                                                            isValid={touched.phone && !errors.phone}
                                                            isInvalid={!!errors.phone}
                                                        />
                                                        <Form.Control.Feedback>Looks good!</Form.Control.Feedback>
                                                        <Form.Control.Feedback type="invalid">
                                                            {errors.phone}
                                                        </Form.Control.Feedback>
                                                    </Form.Group>
                                                    <Button
                                                        variant="outline-primary"
                                                        type="submit"
                                                        block={true}
                                                        disabled={isSubmitting}
                                                    >
                                                        Update Profile
                                                    </Button>
                                                </Form>
                                            )}
                                        />
                                    </CenterContainer>
                                </Card.Body>
                            </Card>
                        </CardWrapper>
                    </Col>
                </Row>
            </Container>
        );
    }
}

const CenterContainer = styled.div`
    text-align: center;
`;

const CardWrapper = styled.div`
    margin-top: 5rem;
`;

const mapStateToProps = (state: ApplicationState) => ({
    expertUser: state.server.expertUser,
});
export default withRouter(
    connect(
        mapStateToProps,
        {
            fetchExpertUserAction,
            expertProfileUpdateAction,
        },
    )(ExpertProfileUpdate),
);
