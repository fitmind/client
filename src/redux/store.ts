import { Store, createStore, applyMiddleware } from 'redux';
import { ApplicationState, storeInitialState, createRootReducer } from './reducers/root.reducer';
import { composeWithDevTools } from 'redux-devtools-extension';
import createSagaMiddleware from 'redux-saga';
import { rootSaga } from './saga/root.saga';
import { History, createBrowserHistory } from 'history';
import { routerMiddleware } from 'connected-react-router';

// We use hash history because this example is going to be hosted statically.
// Normally you would use browser history.
export const history: History = createBrowserHistory();

export const configureStore = (initialState: ApplicationState): Store<ApplicationState> => {
    const sagaMiddleware = createSagaMiddleware();
    const composeEnhancers = composeWithDevTools({});
    let store = createStore(
        createRootReducer(history),
        initialState,
        composeEnhancers(applyMiddleware(routerMiddleware(history), sagaMiddleware)),
    );

    sagaMiddleware.run(rootSaga);
    return store;
};

const store: Store<ApplicationState> = configureStore(storeInitialState);

export default store;
