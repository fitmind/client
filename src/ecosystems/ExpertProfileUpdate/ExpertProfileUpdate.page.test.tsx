import { shallow } from 'enzyme';
import React from 'react';
import { Provider } from 'react-redux';
import { Router } from 'react-router';
import renderer from 'react-test-renderer';
import store, { history } from '../../redux/store';
import ExpertProfileUpdate from './ExpertProfileUpdate.page';

describe('Expert Profile Update page', () => {
    it('renders without crashing', () => {
        shallow(
            <Provider store={store}>
                <Router history={history}>
                    <ExpertProfileUpdate />
                </Router>
            </Provider>,
        );
    });

    it('should match the snapshot', () => {
        const component = renderer.create(
            <Provider store={store}>
                <Router history={history}>
                    <ExpertProfileUpdate />
                </Router>
            </Provider>,
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});
