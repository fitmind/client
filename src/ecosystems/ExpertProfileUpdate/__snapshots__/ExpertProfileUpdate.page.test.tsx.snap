// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expert Profile Update page should match the snapshot 1`] = `
<div
  className="container-fluid"
>
  <div
    className="row"
  >
    <div
      className="col-md-4"
    />
    <div
      className="col-md-4"
    >
      <div
        className="sc-bwzfXH kUNjUS"
      >
        <div
          className="card"
        >
          <h5
            className="card-header"
          >
            Edit your Profile
          </h5>
          <div
            className="card-body"
          >
            <div
              className="sc-bdVaJa kmXoBq"
            >
              <img
                className="rounded"
              />
              <form
                className=""
                noValidate={true}
                onSubmit={[Function]}
              >
                <div
                  className="form-group"
                >
                  <label
                    className="form-label"
                  >
                    First Name
                  </label>
                  <input
                    className="form-control"
                    name="firstName"
                    onBlur={[Function]}
                    onChange={[Function]}
                    placeholder="your name"
                    type="text"
                  />
                  <div
                    className="valid-feedback"
                  >
                    Looks good!
                  </div>
                  <div
                    className="invalid-feedback"
                  />
                </div>
                <div
                  className="form-group"
                >
                  <label
                    className="form-label"
                  >
                    Last Name
                  </label>
                  <input
                    className="form-control"
                    name="lastName"
                    onBlur={[Function]}
                    onChange={[Function]}
                    placeholder="your last name"
                    type="text"
                  />
                  <div
                    className="valid-feedback"
                  >
                    Looks good!
                  </div>
                  <div
                    className="invalid-feedback"
                  />
                </div>
                <div
                  className="form-group"
                >
                  <label
                    className="form-label"
                  >
                    Your Description
                  </label>
                  <textarea
                    className="form-control"
                    name="description"
                    onBlur={[Function]}
                    onChange={[Function]}
                    rows={4}
                    type="text"
                  />
                  <div
                    className="valid-feedback"
                  >
                    Looks good!
                  </div>
                  <div
                    className="invalid-feedback"
                  />
                </div>
                <div
                  className="form-group"
                >
                  <label
                    className="form-label"
                  >
                    What type of coaching are you looking at?
                  </label>
                  <select
                    className="form-control"
                    multiple={true}
                    name="interestedInExpertiseAreas"
                    onBlur={[Function]}
                    onChange={[Function]}
                  >
                     
                    <option
                      selected={false}
                      value="YOGA_COACH"
                    >
                      Yoga Teacher
                    </option>
                    <option
                      selected={false}
                      value="PERSONAL_COACH"
                    >
                      Personal Trainer
                    </option>
                    <option
                      selected={false}
                      value="NUTRITIONIST"
                    >
                      Nutritionist
                    </option>
                    <option
                      selected={false}
                      value="LIFE_COACH"
                    >
                      Life Coach
                    </option>
                  </select>
                </div>
                <div
                  className="form-group"
                >
                  <label
                    className="form-label"
                  >
                    Phone
                  </label>
                  <input
                    className="form-control"
                    name="phone"
                    onBlur={[Function]}
                    onChange={[Function]}
                    placeholder="Enter phone"
                    type="text"
                  />
                  <div
                    className="valid-feedback"
                  >
                    Looks good!
                  </div>
                  <div
                    className="invalid-feedback"
                  />
                </div>
                <button
                  className="btn btn-outline-primary btn-block"
                  disabled={false}
                  type="submit"
                >
                  Update Profile
                </button>
              </form>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
